zap_otlp_encoder/zap_otlp_encoder.go

```diff
@@ -0,0 +1,237 @@
+	enc.addKeyVal(key, &v1.AnyValue{Value: &v1.AnyValue_IntValue{IntValue: int64(val)}})
+}
+
+func (enc *otlpEncoder) AddTime(key string, val time.Time) {
+	// not required
+}
+
+func (enc *otlpEncoder) AddReflected(key string, obj interface{}) error {
+	return nil
+}
+
+func (enc *otlpEncoder) OpenNamespace(key string) {
+	// todo: implement this later
+}
+
+func (enc *otlpEncoder) AddInt(k string, v int)         {}
+func (enc *otlpEncoder) AddInt32(k string, v int32)     {}
+func (enc *otlpEncoder) AddInt16(k string, v int16)     {}
+func (enc *otlpEncoder) AddInt8(k string, v int8)       {}
+func (enc *otlpEncoder) AddUint(k string, v uint)       {}
+func (enc *otlpEncoder) AddUint32(k string, v uint32)   {}
+func (enc *otlpEncoder) AddUint16(k string, v uint16)   {}
+func (enc *otlpEncoder) AddUint8(k string, v uint8)     {}
+func (enc *otlpEncoder) AddUintptr(k string, v uintptr) {}
+
+func (enc *otlpEncoder) Clone() zapcore.Encoder {
+	clone := enc.clone()
+	clone.buf.Write(enc.buf.Bytes())
+	clone.log = &lpb.LogRecord{}
+	*clone.log = *enc.log
+	return clone
+}
+
+func (enc *otlpEncoder) clone() *otlpEncoder {
+	clone := getOTLPEncoder()
+	clone.EncoderConfig = enc.EncoderConfig
+	clone.spaced = enc.spaced
+	clone.openNamespaces = enc.openNamespaces
+	clone.buf = bufferPool.Get()
+	clone.log = &lpb.LogRecord{}
+	return clone
+}
+
+var levelMap = map[string]lpb.SeverityNumber{
+	"trace": lpb.SeverityNumber_SEVERITY_NUMBER_TRACE,
+	"debug": lpb.SeverityNumber_SEVERITY_NUMBER_DEBUG,
+	"info":  lpb.SeverityNumber_SEVERITY_NUMBER_DEBUG,
+	"warn":  lpb.SeverityNumber_SEVERITY_NUMBER_WARN,
+	"error": lpb.SeverityNumber_SEVERITY_NUMBER_ERROR,
+	"fatal": lpb.SeverityNumber_SEVERITY_NUMBER_FATAL,
+}
+
+func (enc *otlpEncoder) EncodeEntry(ent zapcore.Entry, fields []zapcore.Field) (*buffer.Buffer, error) {
+	final := enc.clone()
+	logger := ""
+	if final.LevelKey != "" && final.EncodeLevel != nil {
+		final.log.SeverityNumber = levelMap[ent.Level.String()]
+		final.log.SeverityText = lpb.SeverityNumber_name[int32(final.log.SeverityNumber)]
+	}
+	if final.TimeKey != "" {
+		final.log.TimeUnixNano = uint64(ent.Time.UnixNano())
+	}
+	if ent.LoggerName != "" {
+		// This will be converted to instrumentation scope later
+		logger = ent.LoggerName
+	}
+	if ent.Caller.Defined {
+		if final.CallerKey != "" {
+			final.addKeyVal(final.CallerKey, &v1.AnyValue{Value: &v1.AnyValue_StringValue{StringValue: ent.Caller.String()}})
+		}
+		if final.FunctionKey != "" {
+			final.addKeyVal(final.FunctionKey, &v1.AnyValue{Value: &v1.AnyValue_StringValue{StringValue: ent.Caller.Function}})
+		}
+	}
+	if final.MessageKey != "" {
+		final.log.Body = &v1.AnyValue{
+			Value: &v1.AnyValue_StringValue{
+				StringValue: ent.Message,
+			},
+		}
+	}
+
+	addFields(final, fields)
+	data, err := proto.Marshal(final.log)
+	if err != nil {
+		return nil, err
+	}
+
+	final.buf.AppendString(logger + "#SIGNOZ#" + string(data))
```

Comment:
We can still keep the name as a part of the attributes here instead of this separator logic. And in the transporter/sync program just delete the attribute and map it to the instrumentation scope.

Reply:
Attribute is an array which will result in searching and removing the attribute, I feel this one is a simpler implementation compared to that, but happy to hear your comments.

Reply:
Since we want to advocate this to a larger audience, I suggested the attributes approach, given that attributes is usually small-size array. And want to avoid buggy cases where separator keyword appears in `data` rare but still there is a chance. I don't have a strong opinion here.